### PR TITLE
[temporary] additionally builds graph and test into esm

### PR DESCRIPTION
**Pull request by ZempTime**
Attempt at addressing request for PR from here: https://github.com/davidkpiano/xstate/issues/1025

I'm not sure I did this right, so please let me know any changes you'd like. Specifically, here are the areas I think I still need to address:

- [x] Should I add `.web` entries for the `package.json`'s `files` declaration? (Or conversely, remove the web build targets from rollup config?)
- [x] Verify builds (see below)
- [x] Changeset

### Verifying Builds
Whenever I try to `yarn build` (or more specifically, `tsc`), I run into:
```
src/graph.ts:8:3 - error TS2724: Module ‘“../../../../../../../Users/zempelc/code/node_modules/xstate/lib”’ has no exported member ‘AnyEventObject’. Did you mean ‘DoneEventObject’?
8   AnyEventObject
    ~~~~~~~~~~~~~~
Found 1 error.
error Command failed with exit code 2.
```

`AnyEventObject`, however, is _definitely_ there and exported from said directory:

Screen Shot 2020-02-28 at 10 19 51 AM

Which leads me to think I have a misconfiguration somewhere.

When I add `xstate` as a dep (not a peerDep or devDep), I'm able to build.

### CODE REVIEW

**Comment by Andarist** [comment]
Ill get to reviewing this today/tomorrow, thanks for the PR

File: packages/xstate-test/package.json [hunk 1]
```diff
@@ -17,14 +17,17 @@
...
   },
   "sideEffects": false,
   "files": [
     "lib/**/*.js",
-    "lib/**/*.d.ts"
+    "lib/**/*.d.ts",
+    "es/**/*.js",
+    "es/**/*.d.ts"
   ],
   "repository": {
```
Comment (ZempTime) [inline packages/xstate-test/package.json]: Add/update to:

```
    "build": "tsc && rollup -c && npm run annotate:es",
    "annotate:es": "babel es --out-dir es --no-babelrc --plugins annotate-pure-calls",
```

File: packages/xstate-graph/package.json [hunk 2]
```diff
@@ -14,19 +14,23 @@
...
   ],
   "repository": {
     "type": "git",
     "url": "git+ssh://git@github.com/davidkpiano/xstate.git"
   },
   "scripts": {
     "clean": "rm -rf dist lib tsconfig.tsbuildinfo",
-    "build": "tsc",
+    "build": "tsc && rollup -c && npm run annotate:es",
+    "annotate:es": "babel es --out-dir es --no-babelrc --plugins annotate-pure-calls",
```
Comment (Andarist) [inline packages/xstate-graph/package.json]: let's not do this right now, this has been removed on the `next` branch

File: packages/xstate-graph/package.json [hunk 3]
```diff
@@ -14,19 +14,23 @@
   "homepage": "https://github.com/davidkpiano/xstate/tree/master/packages/xstate-test#readme",
   "license": "MIT",
   "main": "lib/index.js",
+  "module": "es/index.js",
```
Comment (Andarist) [inline packages/xstate-graph/package.json]: @davidkpiano what's the status of `@xstate/graph`? it has not been included in [`publishablePackages`](https://github.com/davidkpiano/xstate/blob/6bda76dec5b5e08b7da432a2b3c35ef9cc44f8e7/.github/actions/change-release-intent/main.js#L26), but i don't recall why. 

When I look at this now it seems that we should just add it to that set of `publishablePackages` because it's a dependency of `@xstate/test` and is already published and on the v1.

File: packages/xstate-graph/rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }),
+  createUmdConfig({
+    input: 'src/index.ts',
+    output: {
+      file: 'dist/xstate-graph.js',
+      format: 'umd',
+      name: 'XStateGraph'
+    }
+  }),
+  createUmdConfig({
```
Comment (Andarist) [inline packages/xstate-graph/rollup.config.js]: let's remove those `.web` builds for now - only `xstate` has it and I'm still unsure how people actually utilize it, or rather - how they would utilize a build like this which has actual external dependencies (xstate has no dependencies, but this one depends on xstate)

File: packages/xstate-graph/package.json [hunk 5]
```diff
@@ -37,6 +41,15 @@
     "xstate": "^4.7.0"
   },
   "devDependencies": {
+    "@babel/cli": "^7.6.0",
```
Comment (Andarist) [inline packages/xstate-graph/package.json]: Seems like you have added those manually and you have not sorted those dependencies correctly.

File: packages/xstate-graph/package.json (hunk 3 above)
Comment (davidkpiano) [reply packages/xstate-graph/package.json]: It should be - it is published.

**Comment by ZempTime** [comment]
I understand your feeling that this is seems like bad timing/inefficient effort.

For context, while this is rather inefficient dev-wise, es modules immediately enables ~10+ other devs to start digging into `test` and `graph` (this will happen soon as it's there). So I'd like to reinforce, this isn't wasted effort. You're directly helping people be able to use this over the short-term. Additionally, I'll be sharing the PoC with some other communities on top of that, so I'm taking a little responsibility for the very first things they'd do, which is install the deps and look for the esm's.

That said, I'm not married to the pr. In the process of doing this, I figured out how to manually build the packages I needed enough to advance with my PoC, and I could totally compensate for installation issues with communication. I opened the original issue to surface the troubles I ran into, and totally open to whatever way you'd wanna process that. Next question on my mind is: when will next get merged in (thus addressing esm need)? But don't answer that. I never ever wanna ask open-source devs (or any, ideally) for time estimates on principle. That's the question it's up to me to figure out.

To answer that for myself, assuming [this](https://github.com/davidkpiano/xstate/issues?q=is%3Aopen+is%3Aissue+milestone%3A5.0) is the right place to look at, might be a little bit.

tl;dr - This PR will immediately produce some adoption/exposure for you, so hopefully knowing it's not wasted effort will make it feel less bad. That said, I'm happy to rescind this if you wanna change your mind.

I'll try and have this ready for you by tomorrow!

**Comment by davidkpiano** [comment]
@zemptime This will still be useful for at least a couple months, so thank you! I think it would still be good to merge in for now.

File: packages/xstate-graph/package.json (hunk 5 above)
Comment (ZempTime) [reply packages/xstate-graph/package.json]: I'm not sure what this means. I tried:

Screen Shot 2020-03-02 at 12 37 07 PM

Would this mean I could
`yarn manypkg run @xstate/graph start`
and not need these deps in this package? I'll start trying things

Comment (ZempTime) [reply packages/xstate-graph/package.json]: Unless you meant alphabetically. Just `:sort`'d them in case

**Comment by ZempTime** [comment]
Awesome. Pushed up some changes to hopefully address the remaining comments. Please note:

* added `es` dir to `yarn clean` (forgot it)
* manually confirmed working builds in both places

Still remaining: 
- [ ] Unsure how to address changesets but happy to do so with more direction
- [ ] _Hope_ I addressed the sorted dependencies?

**Comment by ZempTime** [comment]
Final note, idk if this will be a big deal, this new build will produce

`packages/xstate-test/es/slimChalk.js`
```javascript
import chalk from 'chalk';

function slimChalk(color, str) {
    return chalk[color](str);
}

export default slimChalk;
```
I'm not certain that importing chalk from chalk will work, as that doesn't export esm either

File: packages/xstate-graph/package.json (hunk 5 above)
Comment (Andarist) [reply packages/xstate-graph/package.json]: Sorry for not mentioning that - but yes, I was referring to the reported problem from `manypkg`. Glad you have found how to fix this using that tool 🎉

**Comment by Andarist** [comment]
> This PR will immediately produce some adoption/exposure for you, so hopefully knowing it's not wasted effort will make it feel less bad. That said, I'm happy to rescind this if you wanna change your mind.

I wasn't implying that we shouldn't proceed with that - it was just a general remark that we have solved this already, just on a different branch. As @davidkpiano said - this will still be useful for the v4 release line for a few more months.

> Additionally, I'll be sharing the PoC with some other communities on top of that, so I'm taking a little responsibility for the very first things they'd do, which is install the deps and look for the esm's.

Out of curiosity - how do you plan to consume those ESM builds? You still won't be able to load them with correctly with Node and web bundlers (such as webpack) are capable of loading CJS modules.

> Next question on my mind is: when will next get merged in (thus addressing esm need)? But don't answer that. I never ever wanna ask open-source devs (or any, ideally) for time estimates on principle.

❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ 

> I'm not certain that importing chalk from chalk will work, as that doesn't export esm either

Well, this really depends on how exactly do you plan to load those files 😅 As mentioned - would love to learn how do you plan to consume them.

**Comment by ZempTime** [comment]
On where I'll be consuming ESM builds:

* immediate use is [here](https://github.com/ZempTime/dwight-job-application/blob/master/test/dwight-job-application.test.js#L3) where I'm noobing my way through understanding how to do model-based testing
* More generalized from my messy repo there, the gold standard is Open WC's [development](https://open-wc.org/developing/#going-buildless) and [building](https://open-wc.org/building/#browser-standards) guidelines.
* At work in a custom webpack build/monorepo setup (which I hope to convert to open-wc based at some point)

I need to wrap my head around how to actually implement model-based testing in what I'd loosely label "my ecosystem," and we trend a strange bifurcated mixture of absolute bleeding edge greenfield <--> enterprise needing to deal with every situation ever. Generally, things have settled on "develop using ESM's then build varieties of outputs depending on targets/situation." This is the reason why I'm interested in the `es/` outputs.

I personally see model-based testing as an approach really worth investing into. This is totally personal-opinion territory, but I see it as the layer we (the engineering team at work) provides around reused pieces of ui work (think, micro-frontends or composable feature-based frontends) so other business units/clients can consume either our ui 'business logic' layer, and/or our presentation layer. PLUS, have them compose nicely/be self-documenting. We can provide 'spec' surfaces if other teams wanna implement known functionality surfaces using different tech.

I also think it'd be great to model our deployments and drive testing of those through a model-based approach where the pathways are generated by an xstate model, and we move on those manually. I'm not sure what kind of setup xstate will be getting consumed into in that case once this notion moves off of me writing it manually on my computer.

**Comment by ZempTime** [comment]
Any updates on this? I'll be redoubling on model-based testing, soon. 🤗

**Comment by Andarist** [comment]
@ZempTime sorry, I was under impression that you were about making some changes to the PR. It seems like not and the PR is green - gonna review this again soon.

**Comment by Andarist** [comment]
@ZempTime, in the end, I've ended up using plain `tsc` to generate those directories as it's a fairly short and simple change. 

This might not resolve your issue entirely though, because of `chalk` being used in `@xstate/test`. In bundlers, this works thanks to ["browser redirects"](https://github.com/davidkpiano/xstate/blob/6e78b3bcbea59939d06cf71fc7e0e59cadc56b0b/packages/xstate-test/package.json#L22-L25) in package.json, but I'm not sure if your setup is capable of something like this.
